### PR TITLE
Sol 58057 empty payload

### PR DESCRIPTION
**Pull request by carolmorneau**

### CODE REVIEW

File: solace-spring-cloud-stream-binder/solace-spring-cloud-stream-binder-core/src/main/java/com/solace/spring/cloud/stream/binder/util/XMLMessageMapper.java [hunk 1]
```diff
@@ -207,6 +214,11 @@ public XMLMessage map(Message<?> message, Collection<String> excludedHeaders, bo
 				.setHeaderIfAbsent(MessageHeaders.CONTENT_TYPE, xmlMessage.getHTTPContentType())
 				.setHeaderIfAbsent(IntegrationMessageHeaderAccessor.DELIVERY_ATTEMPT, new AtomicInteger(0));
 
+		if (isNullPayload) {
+			logger.info("Null payload detected, setting Spring header " + SolaceBinderHeaders.NULL_PAYLOAD);
```
Comment (Nephery) [inline solace-spring-cloud-stream-binder/solace-spring-cloud-stream-binder-core/src/main/java/com/solace/spring/cloud/stream/binder/util/XMLMessageMapper.java:218]: Remove this log or change it to a DEBUG log (wrap it in a `logger.isDebugEnabled()`), or a TRACE log (wrap it in a `logger.isTraceEnabled()`). Logging on each message will bog down performance.

File: solace-spring-cloud-stream-binder/solace-spring-cloud-stream-binder-core/src/test/java/com/solace/spring/cloud/stream/binder/util/XMLMessageMapperTest.java [hunk 2]
```diff
@@ -363,6 +363,8 @@ public void testMapSpringMessageToXMLMessage_NonWriteableSolaceProperties() thro
 				case SolaceBinderHeaders.CONFIRM_CORRELATION:
 					assertNull(xmlMessage.getProperties().get(header.getKey()));
 					break;
+				case SolaceBinderHeaders.NULL_PAYLOAD:
+					break; //Nothing to check as this property is not set on the xmLMessage
```
Comment (Nephery) [inline solace-spring-cloud-stream-binder/solace-spring-cloud-stream-binder-core/src/test/java/com/solace/spring/cloud/stream/binder/util/XMLMessageMapperTest.java:367]: Need to check that the `xmlMessage.getProperties()` is `null`. The purpose of this test is ensure that if an application sets a non-writable header on their outbound Spring message, that it is ignored and doesn't get propagated to the `xmlMessage.getProperties()`.

File: solace-spring-cloud-stream-binder/solace-spring-cloud-stream-binder-core/src/test/java/com/solace/spring/cloud/stream/binder/util/XMLMessageMapperTest.java [hunk 3]
```diff
@@ -1071,7 +1076,14 @@ public void testMapXMLMessageToSpringMessage_ReadLocalSolaceProperties() throws
 		Mockito.verify(xmlMessageMapper).map(xmlMessage, acknowledgmentCallback, false);
 
 		for (Map.Entry<String, ? extends HeaderMeta<?>> header : readableLocalHeaders) {
-			fail(String.format("no test for header %s", header.getKey()));
+			Object actualValue = springMessage.getHeaders().get(header.getKey());
+			switch (header.getKey()) {
+				case SolaceBinderHeaders.NULL_PAYLOAD:
```
Comment (Nephery) [inline solace-spring-cloud-stream-binder/solace-spring-cloud-stream-binder-core/src/test/java/com/solace/spring/cloud/stream/binder/util/XMLMessageMapperTest.java:1081]: Can't add a comment there for some reason, but on line `1060`, can you uncomment the `assertNotEquals("Test header set was empty", 0, readableLocalHeaders.size());`? Now that we have actual local-scoped readable headers, we can actually start running that assertion. 😆

**Review by Nephery — CHANGES_REQUESTED** [review]

File: solace-spring-cloud-starters/solace-spring-cloud-stream-starter/README.adoc [hunk 4]
```diff
@@ -724,6 +730,12 @@ public class MyMessageBuilder {
 
 Also, this header is cleared by the message's producer before it is sent off to the message broker. So you should attach the target destination to your message payload if you want to get that information on the consumer-side.
 
+== Empty Payload VS Null Payload
+
+Spring messages can't contain null payloads, however, message handlers can differentiate between null payloads and empty payloads by looking at the `solace_scst_nullPayload` header. The binder adds the `solace_scst_nullPayload` header when a Solace message with null payload is consumed from the wire. When that is the case, the binder sets the Spring message's payload to a null equivalent payload. Null equivalent payloads are one of the following: empty byte[], empty String, empty SDTMap, or empty SDTStream.
+
+Note that applications can't differentiate between null payloads and empty payloads when consuming BytesMessage or XMLContentMessage from the wire. This is because Solace always converts empty payloads to null payloads when those message types are published.
```
[the remark this answers is not in the document]
Comment (carolmorneau) [reply solace-spring-cloud-starters/solace-spring-cloud-stream-starter/README.adoc]: Done

File: solace-spring-cloud-stream-binder/solace-spring-cloud-stream-binder-core/src/main/java/com/solace/spring/cloud/stream/binder/util/XMLMessageMapper.java (hunk 1 above)
Comment (carolmorneau) [reply solace-spring-cloud-stream-binder/solace-spring-cloud-stream-binder-core/src/main/java/com/solace/spring/cloud/stream/binder/util/XMLMessageMapper.java]: Done

File: solace-spring-cloud-stream-binder/solace-spring-cloud-stream-binder-core/src/test/java/com/solace/spring/cloud/stream/binder/util/XMLMessageMapperTest.java (hunk 2 above)
Comment (carolmorneau) [reply solace-spring-cloud-stream-binder/solace-spring-cloud-stream-binder-core/src/test/java/com/solace/spring/cloud/stream/binder/util/XMLMessageMapperTest.java]: Right, thanks
Done

File: solace-spring-cloud-stream-binder/solace-spring-cloud-stream-binder-core/src/test/java/com/solace/spring/cloud/stream/binder/util/XMLMessageMapperTest.java (hunk 3 above)
Comment (carolmorneau) [reply solace-spring-cloud-stream-binder/solace-spring-cloud-stream-binder-core/src/test/java/com/solace/spring/cloud/stream/binder/util/XMLMessageMapperTest.java:1078]: You were forward thinking when you wrote that ;)

**Review by Nephery — APPROVED** [review]

**Pull request merged**